packages/backend-common/src/reading/tree/ArchiveResponse.ts

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import os from 'os';
+import tar, { Parse, ParseStream, ReadEntry } from 'tar';
+import path from 'path';
+import fs from 'fs-extra';
+import { Readable, pipeline as pipelineCb } from 'stream';
+import { promisify } from 'util';
+import concatStream from 'concat-stream';
+import { ReadTreeResponse, File, ReadTreeResponseDirOptions } from '../types';
+
+// Tar types for `Parse` is not a proper constructor, but it should be
+const TarParseStream = (Parse as unknown) as { new (): ParseStream };
+
+const pipeline = promisify(pipelineCb);
+
+/**
+ * Wraps a tar archive stream into a tree response reader.
+ */
+export class ArchiveResponse implements ReadTreeResponse {
+  private read = false;
+
+  constructor(
+    private readonly stream: Readable,
+    private readonly subPath: string,
+    private readonly workDir: string = os.tmpdir(),
+    private readonly filter?: (path: string) => boolean,
+  ) {
+    if (subPath) {
+      if (!subPath.endsWith('/')) {
+        this.subPath += '/';
+      }
+      if (subPath.startsWith('/')) {
+        throw new TypeError(
+          `ArchiveResponse subPath must not start with a /, got '${subPath}'`,
+        );
+      }
+    }
+  }
+
+  // Make sure the input stream is only read once
+  private onlyOnce() {
+    if (this.read) {
+      throw new Error('Response has already been read');
+    }
+    this.read = true;
+  }
+
+  async files(): Promise<File[]> {
+    this.onlyOnce();
+
+    const files: File[] = [];
+    const parser = new TarParseStream();
+
+    parser.on('entry', (entry: ReadEntry & Readable) => {
+      if (entry.type === 'Directory') {
+        entry.resume();
+        return;
+      }
+
+      if (this.subPath) {
+        if (!entry.path.startsWith(this.subPath)) {
+          entry.resume();
+          return;
+        }
+      }
+
+      const path = this.subPath
+        ? entry.path.replace(this.subPath, '')
+        : entry.path;
+      if (this.filter) {
+        if (!this.filter(path)) {
+          entry.resume();
+          return;
+        }
+      }
+
+      const content = new Promise<Buffer>(async resolve => {
+        await pipeline(entry, concatStream(resolve));
+      });
```

Comment:
I do understand it right that all files get loaded eagerly into memory, correct? Not saying it can/shall be fixed, just checking. We can't consume the stream twice after all, and barring using a temp folder this is what we can do

Reply:
Yeah tried to avoid it, but not really possible to do since we're returning a list of files and therefore need to process the entire stream before we can return. We could switch to returning an async iterator of file entries or something like that though.

Reply:
Only time it's wasteful is of you just want to list the file structure but don't care about the contents, and that's not something we should worry too much about...

Reply:
Yeah my main concern was to only do this when `files()` is called. If you're in a situation where you may be working with large amount of data you should use one of the other alternatives.

Reply:
Right.

And we could add a `filesMeta()` if we wanted to, some day, for just listing the contents but not retain data, if it were of interest to somebody.

Speaking of which, maybe the `File` type could contain `meta: FileMeta` and `data` where `FileMeta` could be extended later if we wanted with other (optional) things like gid, uid, flags, type (if symlink) etc? Or is that bikeshedding :)